### PR TITLE
[API stabilization] Encoder and Decoder

### PR DESCRIPTION
**Pull request by qwwdfsad**
* Deprecate potentially dangerous extensions
    * Rename context to serializersModule to align with modules rework
    * Remove decode/encodeUnit
    * Rename READ_DONE to DECODE_DONE

### CODE REVIEW

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt [hunk 1]
```diff
@@ -169,7 +168,7 @@ public class Cbor(
 
         override fun skipBeginToken() = setSize(decoder.startArray())
 
-        override fun decodeElementIndex(descriptor: SerialDescriptor) = if (!finiteMode && decoder.isEnd() || (finiteMode && ind >= size)) READ_DONE else ind++
+        override fun decodeElementIndex(descriptor: SerialDescriptor) = if (!finiteMode && decoder.isEnd() || (finiteMode && ind >= size)) CompositeDecoder.DECODE_DONE else ind++
```
Comment (sandwwraith) [inline formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:171]: why not import this constant?

File: formats/config/src/main/kotlin/kotlinx/serialization/config/ConfigReader.kt [hunk 2]
```diff
@@ -84,7 +83,7 @@ public class ConfigParser(
                     return ind
                 }
             }
-            return READ_DONE
+            return CompositeDecoder.DECODE_DONE
```
Comment (sandwwraith) [inline formats/config/src/main/kotlin/kotlinx/serialization/config/ConfigReader.kt:86]: same

File: runtime/api/kotlinx-serialization-runtime.api [hunk 3]
```diff
@@ -29,15 +30,15 @@ public abstract interface class kotlinx/serialization/CompositeDecoder {
...
 	public abstract fun endStructure (Lkotlinx/serialization/SerialDescriptor;)V
-	public abstract fun getContext ()Lkotlinx/serialization/modules/SerialModule;
+	public abstract fun getSerializersModule ()Lkotlinx/serialization/modules/SerialModule;
 	public abstract fun getUpdateMode ()Lkotlinx/serialization/UpdateMode;
 	public abstract fun updateNullableSerializableElement (Lkotlinx/serialization/SerialDescriptor;ILkotlinx/serialization/DeserializationStrategy;Ljava/lang/Object;)Ljava/lang/Object;
 	public abstract fun updateSerializableElement (Lkotlinx/serialization/SerialDescriptor;ILkotlinx/serialization/DeserializationStrategy;Ljava/lang/Object;)Ljava/lang/Object;
 }
 
 public final class kotlinx/serialization/CompositeDecoder$Companion {
+	public static final field DECODE_DONE I
```
Comment (sandwwraith) [inline runtime/api/kotlinx-serialization-runtime.api:41]: btw, why ABI duplicated in both interface and companion?

Comment (qwwdfsad) [reply runtime/api/kotlinx-serialization-runtime.api:41]: Unfortunately, this is our compilation strategy for companions in  interfaces

File: formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt (hunk 1 above)
Comment (qwwdfsad) [reply formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:171]: This is an automatic replacement.
Replacing just with `ReplaceWith("DECODE_DONE ")` isn't working in IDEA in corner cases.
Fixed

**Pull request merged**